File: src/AST.hs
-- | Abstract Syntax Tree for the Homun language.
-- Every construct maps directly to a Rust construct.
module AST where

type Name = String

-- | A top-level Homun program is a sequence of statements
type Program = [Stmt]

-- ─────────────────────────────────────────
-- Statements
-- ─────────────────────────────────────────

data Stmt
  = SBind     Name   Expr              -- name := expr
  | SUse      [Name]                   -- use engine::physics::{Vec2}
  | SStructDef Name [FieldDef]         -- Player := struct { ... }
  | SEnumDef  Name [VariantDef]        -- Direction := enum { North, South(int) }
  | SExprStmt Expr                     -- bare expression (e.g. function call side-effect)
  deriving (Show)

data FieldDef   = FieldDef   Name (Maybe TypeExpr)  deriving (Show)
data VariantDef = VariantDef Name (Maybe TypeExpr)  deriving (Show)

-- ─────────────────────────────────────────
-- Expressions
-- ─────────────────────────────────────────

data Expr
  -- Literals
  = EInt    Integer
  | EFloat  Double
  | EBool   Bool
  | EString String          -- may contain ${...} interpolations
  | ENone

  -- Variables / field access
  | EVar    Name
  | EField  Expr Name       -- expr.field
  | EIndex  Expr Expr       -- expr[expr]
  | ESlice  Expr (Maybe Expr) (Maybe Expr) (Maybe Expr)   -- expr[a:b:c]

  -- Collections
  | EList   [Expr]                    -- @[...]
  | EDict   [(Expr, Expr)]            -- @{k: v, ...}
  | ESet    [Expr]                    -- @(...)
  | ETuple  [Expr]                    -- (a, b, c)

  -- Struct literal
  | EStruct (Maybe Name) [(Name, Expr)]   -- Player { name: "Aria", hp: 100 }

  -- Operators
  | EBinOp  BinOp Expr Expr
  | EUnOp   UnOp  Expr

  -- Pipe: lhs | f(args)
  | EPipe   Expr Expr                 -- desugars at codegen time

  -- Lambda
  | ELambda [Param] (Maybe TypeExpr) (Maybe TypeExpr) [Stmt] Expr
  --         params  ret-type         void-marker      stmts  final-expr

  -- Application
  | ECall   Expr [Expr]

  -- Control flow as expressions
  | EIf     Expr [Stmt] Expr (Maybe ([Stmt], Expr))  -- if cond do { stmts; expr } else { stmts; expr }
  | EMatch  Expr [MatchArm]

  -- Loops (produce unit / break value)
  | EFor    Name Expr [Stmt] (Maybe Expr)            -- for x in iter do { ... }
  | EWhile  Expr [Stmt] (Maybe Expr)                 -- while cond do { ... }

  -- Break / continue
  | EBreak  (Maybe Expr)
  | EContinue

  -- RON  load/save
  | ELoadRon  Expr TypeExpr   -- load_ron("...") as Type
  | ESaveRon  Expr Expr       -- save_ron(data, "...")

  -- Destructuring bind (used inside block stmts, here for completeness)
  | ERange  (Maybe Expr) (Maybe Expr) (Maybe Expr)   -- range(a,b,c)

  deriving (Show)

-- ─────────────────────────────────────────
-- Match arm
-- ─────────────────────────────────────────

data MatchArm = MatchArm
  { armPat   :: Pat
  , armGuard :: Maybe Expr     -- _ if cond =>
  , armBody  :: Expr
  } deriving (Show)

data Pat
  = PWild                              -- _
  | PVar  Name                         -- variable binding
  | PLit  Expr                         -- literal
  | PTuple [Pat]                       -- (a, b, c)
  | PEnum Name (Maybe Pat)             -- Direction.North  /  Element.Fire(p)
  | PNone                              -- none
  deriving (Show)

-- ─────────────────────────────────────────
-- Types
-- ─────────────────────────────────────────

data TypeExpr
  = TName   Name            -- int, str, bool, float, PlayerState ...
  | TList   TypeExpr        -- @[T]
  | TDict   TypeExpr TypeExpr
  | TSet    TypeExpr
  | TOption TypeExpr        -- none-able
  | TTuple  [TypeExpr]
  | TVoid                   -- _ (no return value)
  | TInfer                  -- unknown, let Rust figure it out
  deriving (Show)

-- ─────────────────────────────────────────
-- Operators
-- ─────────────────────────────────────────

data BinOp
  = OpAdd | OpSub | OpMul | OpDiv | OpMod
  | OpEq  | OpNeq | OpLt | OpGt | OpLe | OpGe
  | OpAnd | OpOr
  | OpIn  | OpNotIn
  deriving (Show, Eq)

data UnOp
  = OpNot | OpNeg
  deriving (Show, Eq)

-- ─────────────────────────────────────────
-- Lambda parameter
-- ─────────────────────────────────────────

data Param = Param
  { paramName :: Name
  , paramType :: Maybe TypeExpr
  } deriving (Show)

File: src/ast.rs
//! Abstract Syntax Tree for the Homun language.
//! Every construct maps directly to a Rust construct.

pub type Name = String;
pub type Program = Vec<Stmt>;

// ─────────────────────────────────────────
// Statements
// ─────────────────────────────────────────

#[derive(Debug, Clone)]
pub enum Stmt {
    Bind(Name, Expr),
    Use(Vec<Name>),
    StructDef(Name, Vec<FieldDef>),
    EnumDef(Name, Vec<VariantDef>),
    Expression(Expr),
}

#[derive(Debug, Clone)]
pub struct FieldDef {
    pub name: Name,
    pub ty: Option<TypeExpr>,
}

#[derive(Debug, Clone)]
pub struct VariantDef {
    pub name: Name,
    pub payload: Option<TypeExpr>,
}

// ─────────────────────────────────────────
// Expressions
// ─────────────────────────────────────────

#[derive(Debug, Clone)]
pub enum Expr {
    // Literals
    Int(i64),
    Float(f64),
    Bool(bool),
    Str(String),
    None,

    // Variables / field access
    Var(Name),
    Field(Box<Expr>, Name),
    Index(Box<Expr>, Box<Expr>),
    Slice(
        Box<Expr>,
        Option<Box<Expr>>,
        Option<Box<Expr>>,
        Option<Box<Expr>>,
    ),

    // Collections
    List(Vec<Expr>),
    Dict(Vec<(Expr, Expr)>),
    Set(Vec<Expr>),
    Tuple(Vec<Expr>),

    // Struct literal
    Struct(Option<Name>, Vec<(Name, Expr)>),

    // Operators
    BinOp(BinOp, Box<Expr>, Box<Expr>),
    UnOp(UnOp, Box<Expr>),

    // Pipe: lhs | f(args)
    Pipe(Box<Expr>, Box<Expr>),

    // Lambda
    Lambda {
        params: Vec<Param>,
        ret_ty: Option<TypeExpr>,
        void_mark: Option<TypeExpr>,
        stmts: Vec<Stmt>,
        final_expr: Box<Expr>,
    },

    // Application
    Call(Box<Expr>, Vec<Expr>),

    // Control flow as expressions
    If(
        Box<Expr>,
        Vec<Stmt>,
        Box<Expr>,
        Option<(Vec<Stmt>, Box<Expr>)>,
    ),
    Match(Box<Expr>, Vec<MatchArm>),

    // Loops
    For(Name, Box<Expr>, Vec<Stmt>, Option<Box<Expr>>),
    While(Box<Expr>, Vec<Stmt>, Option<Box<Expr>>),

    // Bare block
    Block(Vec<Stmt>, Box<Expr>),

    // Break / continue
    Break(Option<Box<Expr>>),
    Continue,

    // RON load/save
    LoadRon(Box<Expr>, TypeExpr),
    SaveRon(Box<Expr>, Box<Expr>),

    // Range
    Range(Option<Box<Expr>>, Option<Box<Expr>>, Option<Box<Expr>>),
}

// ─────────────────────────────────────────
// Match arm
// ─────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct MatchArm {
    pub pat: Pat,
    pub guard: Option<Expr>,
    pub body: Expr,
}

#[derive(Debug, Clone)]
pub enum Pat {
    Wild,
    Var(Name),
    Lit(Expr),
    Tuple(Vec<Pat>),
    Enum(Name, Option<Box<Pat>>),
    None,
}

// ─────────────────────────────────────────
// Types
// ─────────────────────────────────────────

#[derive(Debug, Clone)]
pub enum TypeExpr {
    Name(Name),
    List(Box<TypeExpr>),
    Dict(Box<TypeExpr>, Box<TypeExpr>),
    Set(Box<TypeExpr>),
    Option(Box<TypeExpr>),
    Tuple(Vec<TypeExpr>),
    Void,
    Infer,
}

// ─────────────────────────────────────────
// Operators
// ─────────────────────────────────────────

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Eq,
    Neq,
    Lt,
    Gt,
    Le,
    Ge,
    And,
    Or,
    In,
    NotIn,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum UnOp {
    Not,
    Neg,
}

// ─────────────────────────────────────────
// Lambda parameter
// ─────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct Param {
    pub name: Name,
    pub ty: Option<TypeExpr>,
}
